### ISSUE TITLE
perf(proxy): What does the proxy do to kill time?

### ISSUE THREAD

**Issue by duesee** [duesee/imap-flow#79]
I see a high CPU load when running the proxy even when there are no events. (I have some ideas what happens but we should investigate.)

**Comment by jakoschiko** [comment]
Good opportunity for testing some profiling tools: https://nnethercote.github.io/perf-book/profiling.html
(but probably it's just some busy waiting on the TCP streams)

**Comment by jakoschiko** [comment]
Okay, you can ignore the flamegraph above. I know what's happening.

Steps for reproducing the issue:

- Start the proxy
- Connect with Thunderbird
- Close Thunderbird
- Observe high CPU with high probability

What's happining? In the happy path we see a log message like this after closing Thunderbird:

```
ERROR service{name="Zoho Mail"}:client{addr=127.0.0.1:52138}: Connection terminated role="c2p" error=Broken pipe (os error 32)
 WARN service{name="Zoho Mail"}:client{addr=127.0.0.1:52138}: done
```

In that case the proxy for the mentioned service/client will be deinitialized and everything is fine. But with a high probability this log message is missing for one of the services. Then the CPU usage goes up and the flamegraph looks like this:

![flamegraph](https://github.com/duesee/imap-flow/assets/44508205/4beaf0a0-8e8c-4eab-a446-e04368d5fa1f)

In this case at least one `ClientFlow` instance is doing busy waiting because of this code:

```rust
let crlf_result = match find_crlf(&self.read_buffer[self.seen_bytes..], self.crlf_relaxed) {
    Some(crlf_result) => crlf_result,
    None => {
        // No full line received yet, more data needed.
        stream.0.read_buf(&mut self.read_buffer).await?;
        return Ok(None);
    }
};
```
The function `read_buf` is returning immediately without reading any bytes.

TLDR:

- If I close the client, there is sometimes no client stream error
- In this case, the proxy continues polling the client stream
- `read_buf` on the client stream returns immediately without reading bytes
- => busy waiting

**Comment by duesee** [comment]
Great analysis!
